parlai/agents/seq2seq/modules.py

```diff
@@ -27,277 +44,291 @@ def pad(tensor, length, dim=0):
+        shared_rnn = self.decoder.rnn if decoder == 'shared' else None
+        self.encoder = RNNEncoder(
+            num_features, embeddingsize, hiddensize,
+            padding_idx=self.NULL_IDX, rnn_class=rnn_class,
+            numlayers=numlayers, dropout=dropout,
+            bidirectional=bidirectional,
             shared_lt=shared_lt, shared_rnn=shared_rnn)
 
-        if self.rank:
-            self.ranker = Ranker(
-                self.decoder,
-                padding_idx=self.NULL_IDX,
-                attn_type=opt['attention'])
-
-        self.beam_log_freq = opt.get('beam_log_freq', 0.0)
-        if self.beam_log_freq > 0.0:
-            self.dict = DictionaryAgent(opt)
-            self.beam_dump_filecnt = 0
-            self.beam_dump_path = opt['model_file'] + '.beam_dump'
-            if not os.path.exists(self.beam_dump_path):
-                os.makedirs(self.beam_dump_path)
-
-    def unbeamize_hidden(self, hidden, beam_size, batch_size):
-        """
-        Creates a view of the hidden where batch axis is collapsed with beam axis,
-        we need to do this for batched beam search, i.e. we emulate bigger mini-batch
-        :param hidden: hidden state of the decoder
-        :param beam_size: beam size, i.e. num of hypothesis
-        :param batch_size: number of samples in the mini batch
-        :return: view of the hidden
-        """
-        if isinstance(hidden, tuple):
-            num_layers = hidden[0].size(0)
-            hidden_size = hidden[0].size(-1)
-            return (hidden[0].view(num_layers, batch_size * beam_size, hidden_size),
-                hidden[1].view(num_layers, batch_size * beam_size, hidden_size))
-        else:  # GRU
-            num_layers = hidden.size(0)
-            hidden_size = hidden.size(-1)
-            return hidden.view(num_layers, batch_size * beam_size, hidden_size)
-
-    def unbeamize_enc_out(self, enc_out, beam_size, batch_size):
-        hidden_size = enc_out.size(-1)
-        return enc_out.view(batch_size * beam_size, -1, hidden_size)
-
-    def forward(self, xs, ys=None, cands=None, cand_indices=None, prev_enc=None,
-                rank_during_training=False, beam_size=1, topk=1):
+        shared_weight = (self.decoder.lt
+                         if lookuptable in ('dec_out', 'all') else None)
+        self.output = OutputLayer(
+            num_features, embeddingsize, hiddensize, dropout=dropout,
+            numsoftmax=numsoftmax, shared_weight=shared_weight)
+
+    def _encode(self, xs, prev_enc=None):
+        """Encode the input or return cached encoder state."""
+        if prev_enc is not None:
+            return prev_enc
+        else:
+            return self.encoder(xs)
+
+    def _starts(self, bsz):
+        """Return bsz start tokens."""
+        return self.START.detach().expand(bsz, 1)
+
+    def _decode_forced(self, ys, encoder_states):
+        """Decode with teacher forcing."""
```

Comment:
I don't understand what this means

Reply:
_decode_forced is putting tokens from the target as the next input to estimate the target score during training or during eval PPL computation

Reply:
"teacher-forcing" always inputs the true token at each time step, vs using the model's predicted